### PR TITLE
Support for block_cache num_shards and other config via option string.

### PR DESCRIPTION
**Pull request by PhaniShekhar**
Problem: Option string accepts only cache_size as parameter for block_cache which is specified as "block_cache=1M".
It doesn't accept other parameters like num_shards etc.

Changes :
1) ParseBlockBasedTableOption in block_based_table_factory is edited to accept cache options in the format "block_cache=<cache_size>:<num_shard_bits>:<strict_capacity_limit>:<high_pri_pool_ratio>".
Options other than cache_size are optional to maintain backward compatibility. The changes are valid for block_cache_compressed as well.
For example, "block_cache=1M:6:true:0.5", "block_cache=1M:6:true", "block_cache=1M:6" and "block_cache=1M" are all valid option strings.

2) Corresponding unit tests are added.

### CODE REVIEW

**Comment by PhaniShekhar** [comment]
Could you please take a look at this PR and let me know if any additional changes are required. Thank you.

File: options/options_test.cc [hunk 1]
```diff
@@ -529,6 +531,65 @@ TEST_F(OptionsTest, GetBlockBasedTableOptionsFromString) {
   ASSERT_EQ(table_opt.cache_index_and_filter_blocks,
             new_opt.cache_index_and_filter_blocks);
   ASSERT_EQ(table_opt.filter_policy, new_opt.filter_policy);
+
+  // make sure default values of block cache are overwritten
+  ASSERT_OK(GetBlockBasedTableOptionsFromString(table_opt,
+             "cache_index_and_filter_blocks=1;"
+             "block_cache=1M:4:true:0.5;"
```
Comment (ajkr) [inline options/options_test.cc]: We have a new pattern for (de)serializing structs in a more consistent/structured way. You can look at how "compaction_options_fifo" is implemented using `ParseStructOptions()`. Do you want to try it out?

Comment (PhaniShekhar) [reply options/options_test.cc]: I updated the PR to use ParseStructOptions. Here are the changes in the commit:
1) Created a LRUCacheOptions structure to hold the required cache
options.
2) A new function NewLRUCache now accepts LRUCacheOptions to create the
cache. This in turn calls the previous NewLRUCache function.
3) lru_cache_options_type_info is added to option_helper.h. This has
type info required for ParseStructOptions.
4) A new option type, kLRUCacheOptions is added and ParseOptionHelper is
modified to parse this new type using ParseStructOptions.
5) ParseBlockBasedTableOption is modified to support block_cache and
block_cache_compressed options in struct like format as
block_cache={capacity=1M;num_shard_bits=4;strict_capacity_limit=true;high_pri_pool_ratio=0.5;}
Backward compatibility is also ensured.
6) Appropriate unit tests are added.

**Comment by PhaniShekhar** [comment]
Could you please take a look at the latest changes using ParseStructOptions. Thank you.

File: cache/lru_cache.cc [hunk 2]
```diff
@@ -513,6 +518,20 @@ size_t LRUCache::TEST_GetLRUSize() {
   return lru_size_of_all_shards;
 }
 
+double LRUCache::TEST_GetHighPriPoolRatio() {
```
Comment (yiwu-arbug) [inline cache/lru_cache.cc]: Remove the TEST_ prefix for both of the methods?

**Comment by sagar0** [comment]
Awesome! This looks pretty good to me, and its great to see that you took advantage of the new ParseStruct...() code. 
One final thing to take care of is that, we recently moved static variables like, lru_cache_options_type_info, from the header into .cc file. So you need to move your static fields too.

File: cache/lru_cache.cc (hunk 2 above)
Comment (PhaniShekhar) [reply cache/lru_cache.cc]: Thank you for the review. I made the required changes and merged with the latest master. Could you please take a look. Thank you.

**Comment by yiwu-arbug** [comment]
@sagar0 good catch.

**Comment by PhaniShekhar** [comment]
I already made the corresponding changes to move "lru_cache_options_type_info" into .cc files as part of the most recent commit "Rework lru_cache_options_type_info to be compatible with work in PR#3178 (Move static variables out of the header file)".

**Pull request closed without merging**